README.md
# Logstash plugins

This repo contains a number of logstash (1.3.3) plugins use by Kerkdienst Gemist B.V.

## Filter: `s3_access_log.rb`

Filter to convert Amazon S3 Server Access Log format to Apache Combined Log Format (CLF) applying some Kerkdienst Gemist specific adjustments along the way.

## Filter: `fingerprint.rb`

Backport (copy) of the logstash 1.4 fingerprint filter (used to generate unique MD5 for unique log lines).

## Input: `s3.rb`

Copy of the logstash 1.3.3 `s3.rb` input module with a single patch applied to allow an empty sincedb file. With this we can specify `sincedb_path = '/dev/null'` to cause the s3 input to always process all log files from a S3 bucket (regardless of the timestamp).

See: [Issue Multiple Files Being Left Unprocessed with Identical Timestamps](https://github.com/logstash-plugins/logstash-input-s3/issues/57)

# Setup

Assuming the logstash (1.3.3) repository has been cloned into `../logstash`.

## Running all specs

To run the specs simply run `make` or `make spec`. This will run all specs in spec/**/*.rb

    $ make

    ............................................
    Finished in 2.56 seconds
    44 examples, 0 failures

## Running a specific spec

    $ HOME=`pwd` ../logstash/bin/logstash rspec spec/filters/icecast_spec.rb
    ...........

    Finished in 11.23 seconds
    11 examples, 0 failures

## Running the logstash agent

To run the agent run `make agent`. This will run the logstash agent with the configuration from conf.d.
This configuration is created by concactenating all files in conf.d together (in sorted order).

	make agent
    # this will run ../logstash/bin/logstash agent --pluginpath lib -f logstash.conf


CHANGELOG.md
## Release 2021-06-11-1435
- [fix] don't use since from sincedb_write but reload from sincedb_read.

## Release 2021-06-11-1348
Merge branch 'feature/fix/large-backlog-issues' into develop
- [enh] if there are more objects available, keep going instead of waiting `interval` seconds
- [fix] limit the number of objects returned to 1000 by default

## Release 2020-12-05-1714
- [enh] switch to date tags for releases
- [fix] join logsources into single string (with comma)
- [fix] join referrer strings with semicolon (;) into single string
- [add] message field containing Apache Common Log format representation of the session
- [add] rubocop and solargraph to have a more pleasant vscode experience

## Release 0.1.1
- [fix] only recalculate bytes for partial content when key extension is .mp3 (MP3 audio content)

## Release 0.1.0
- [fix] ignore image (svg,jpg,png) requests (most importantly poster images)

## Release v0.0.1 - 0.0.8
- Previous releases of logstash-plugins


spec/test_utils.rb
if ENV['COVERAGE']
  require 'simplecov'
  require 'coveralls'

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
    SimpleCov::Formatter::HTMLFormatter,
    Coveralls::SimpleCov::Formatter
  ]
  SimpleCov.start do
    add_filter 'spec/'
    add_filter 'vendor/'
  end
end
require "insist"
require "logstash/agent"
require "logstash/pipeline"
require "logstash/event"
require "logstash/logging"
require "insist"
require "stud/try"

$TESTING = true
if RUBY_VERSION < "1.9.2"
  $stderr.puts "Ruby 1.9.2 or later is required. (You are running: " + RUBY_VERSION + ")"
  $stderr.puts "Options for fixing this: "
  $stderr.puts "  * If doing 'ruby bin/logstash ...' add --1.9 flag to 'ruby'"
  $stderr.puts "  * If doing 'java -jar ... ' add -Djruby.compat.version=RUBY1_9 to java flags"
  raise LoadError
end

$logger = LogStash::Logger.new(STDOUT)
if ENV["TEST_DEBUG"]
  $logger.level = :debug
else
  $logger.level = :error
end

module LogStash
  module RSpec
    def config(configstr)
      let(:config) { configstr }
    end # def config

    def type(default_type)
      let(:default_type) { default_type }
    end
    
    def tags(*default_tags)
      let(:default_tags) { default_tags }
    end

    def fields(default_fields)
      let(:default_fields) { default_fields }
    end

    def sample(sample_event, &block)
      name = sample_event.is_a?(String) ? sample_event : sample_event.to_json
      name = name[0..50] + "..." if name.length > 50

      describe "\"#{name}\"" do
        extend LogStash::RSpec
        let(:pipeline) { LogStash::Pipeline.new(config) }
        let(:event) do
          sample_event = [sample_event] unless sample_event.is_a?(Array)
          next sample_event.collect do |e|
            e = { "message" => e } if e.is_a?(String)
            e['type'] ||= default_type rescue nil # default_type
            default_fields.each do |name, value|
              e[name] ||= value
            end rescue nil # when fields haven't been declared
            next LogStash::Event.new(e)
          end
        end

        let(:results) do
          results = []
          count = 0
          pipeline.instance_eval { @filters.each(&:register) }
          event.each do |e|
            extra = []
            pipeline.filter(e) do |new_event|
              extra << new_event
            end
            results << e if !e.cancelled?
            results += extra.reject(&:cancelled?)
          end

          # TODO(sissel): pipeline flush needs to be implemented.
          #results += pipeline.flush
          next results
        end

        subject { results.length > 1 ? results: results.first }

        it("when processed", &block)
      end
    end # def sample

    def input(&block)
      it "inputs" do
        pipeline = LogStash::Pipeline.new(config)
        queue = Queue.new
        pipeline.instance_eval do 
          @output_func = lambda { |event| queue << event }
        end
        block.call(pipeline, queue)
        pipeline.shutdown
      end
    end # def input

    def agent(&block)
      require "logstash/pipeline"

      it("agent(#{caller[0].gsub(/ .*/, "")}) runs") do
        pipeline = LogStash::Pipeline.new(config)
        pipeline.run
        block.call
      end
    end # def agent

  end # module RSpec
end # module LogStash

class Shiftback
  def initialize(&block)
    @block = block
  end

  def <<(event)
    @block.call(event)
  end
end # class Shiftback


spec/filters/wowza_spec.rb
require "./spec/test_utils"

def validate_wowza_fields

  # fields from WOWZA_ACCESS_LOG
  insist { subject['clientip']    } != nil
  insist { subject['ident']       } != nil
  insist { subject['auth']        } != nil
  insist { subject['timestamp']   } != nil
  insist { subject['verb']        } != nil
  insist { subject['request']     } != nil
  insist { subject['httpversion'] } != nil
  insist { subject['response']    } =~ /\A[0-9]*\z/
  insist { subject['bytes']       } =~ /\A[0-9]*\z/
  insist { subject['referrer']    } != nil
  insist { subject['agent']       } != nil
  insist { subject['duration']    } =~ /\A[0-9]*\z/
end

shared_examples "a valid wowza log parser" do

  sample %(89.99.28.243 - 2cf7e6b063 [19/Sep/2014:06:57:05 +0000] "GET /11723021/2014-09-17-1430.mp3?Signature=BanH0VrdfI%2FXCouvXFXivLBS2PE%3D&Expires=1411116501&AWSAccessKeyId=1VYKRTJ5FFKT5B6F4NR2 HTTP/1.1" 304 1000 "-" "AppleCoreMedia/1.0.0.11B554a (iPad; U; CPU OS 7_0_4 like Mac OS X; nl_nl)" 0) do
    insist { subject['tags'] & %w(_grokparsefailure) } == []
    insist { subject['tags'] }.include? 'wowza_access_log'
    insist { subject['tags'] }.include? 'access_log_timestamp'
    insist { subject['tags'] }.include? 'billable'
    insist { subject['tags'] }.include? 'import' unless nil == subject['path']
    insist { subject['bytes'] } == '1000'
    insist { subject['response'] } == '304'
    insist { subject['timestamp'] } == '19/Sep/2014:06:57:05 +0000'
    insist { subject['logsource'] } == 'video-test'

    validate_wowza_fields
  end

  sample %(89.99.23.76 - - [19/Sep/2014:11:39:31 +0200] "GET /10103060?type=live.mp3 HTTP/1.1" 200 17279640 "http://assets.kerkdienstgemist.nl/player/jw6/6.4.3359/jwplayer.flash.swf" "Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; rv:11.0) like Gecko" 5754) do
    insist { subject['tags'] & %w(_grokparsefailure) } == []
    insist { subject["tags"] }.include? 'wowza_access_log'

    validate_wowza_fields
  end

end

describe "Wowza filters" do

  extend LogStash::RSpec

  describe "with default config" do
    let(:parse_failure_tag) { '_grokparsefailure'}

    fields \
      'type'      => 'syslog',
      'program'   => 'wowza.access.log',
      'logsource' => 'video-test'

    config [ 'filter{',
      File.read("conf.d/06_import_wowza.conf"),
      File.read("conf.d/60_wowza.conf"),
    '}' ].join

    it_behaves_like "a valid wowza log parser"
  end

  describe "with default config" do
    let(:parse_failure_tag) { '_grokparsefailure'}

    fields \
      'type'  => 'import_wowza',
      'path'      => 'video-test.log'

    config [ 'filter{',
      File.read("conf.d/06_import_wowza.conf"),
      File.read("conf.d/60_wowza.conf"),
    '}' ].join

    it_behaves_like "a valid wowza log parser"
  end

end


spec/filters/haproxy_spec.rb
require "./spec/test_utils"

def validate_haproxy_fields
  # fields from haproxy logging
  insist { subject['clientip']             } != nil
  insist { subject['clientport']           } != nil
  insist { subject['timestamp']            } != nil
  insist { subject['frontend_name']        } != nil
  insist { subject['backend_name']         } != nil
  insist { subject['server_name']          } != nil
  insist { subject['time_queue']           } =~ /\A[0-9]*\z/
  insist { subject['time_backend_connect'] } =~ /\A[0-9]*\z/
  insist { subject['time_duration']        } =~ /\A[0-9]*\z/
  insist { subject['bytes']                } =~ /\A[0-9]*\z/
  insist { subject['termination_state']    } != nil
  insist { subject['actconn']              } =~ /\A[0-9]*\z/
  insist { subject['feconn']               } =~ /\A[0-9]*\z/
  insist { subject['beconn']               } =~ /\A[0-9]*\z/
  insist { subject['srvconn']              } =~ /\A[0-9]*\z/
  insist { subject['retries']              } =~ /\A[0-9]*\z/
  insist { subject['srv_queue']            } =~ /\A[0-9]*\z/
end

def validate_haproxy_http_fields
  # fields specific for haproxy http backends
  insist { subject['time_request']             } =~ /\A[0-9]*\z/
  insist { subject['time_backend_response']    } =~ /\A[0-9]*\z/
  insist { subject['response']                 } =~ /\A[0-9]*\z/
  insist { subject['captured_request_cookie']  } != nil
  insist { subject['captured_response_cookie'] } != nil
  insist { subject['http_proto']               } == nil
  insist { subject['http_user']                } == nil
  insist { subject['http_host']                } == nil
  insist { subject['verb']                     } != nil
  insist { subject['request']                  } != nil
  insist { subject['icecast_proto']            } != nil
  insist { subject['http_version']             } != nil
end

shared_examples "a valid haproxy log parser" do

  sample %(46.145.191.22:51472 [19/Sep/2014:14:39:44.501] icecast servers-http/audio-aurum 535/0/1/25/50561 200 19 - - cD-- 44/44/44/7/0 0/0 "SOURCE /10403011 ICE/1.0") do
    insist { subject['tags'] }.include? 'haproxy_icey'
    insist { subject['tags'] }.include? 'haproxy_timestamp'
    insist { subject['tags'] & %w(_grokparsefailure) } == []
    insist { subject['tags'] & %w(billable) } == []

    insist { subject['clientip'] }   == '46.145.191.22'
    insist { subject['clientport'] } == '51472'

    insist { subject['frontend_name'] } == 'icecast'
    insist { subject['backend_name']  } == 'servers-http'
    insist { subject['server_name']   } == 'audio-aurum'

    insist { subject['bytes'] }    == '19'
    insist { subject['response'] } == '200'

    insist { subject['timestamp'] } == '19/Sep/2014:14:39:44.501'
    insist { subject['logsource'] } == 'haproxy-audio-lb-test'

    validate_haproxy_fields
    validate_haproxy_http_fields
  end

  sample %(54.255.254.198:61852 [01/May/2015:11:10:17.902] lb-wowza servers/video-sunfire 0/0/201 296 -- 1/1/1/1/0 0/0) do
    insist { subject['tags'] }.include? 'haproxy_icey'
    insist { subject['tags'] }.include? 'haproxy_timestamp'
    insist { subject['tags'] & %w(_grokparsefailure) } == []
    insist { subject['tags'] & %w(billable) } == []

    insist { subject['clientip'] }   == '54.255.254.198'
    insist { subject['clientport'] } == '61852'

    insist { subject['frontend_name'] } == 'lb-wowza'
    insist { subject['backend_name']  } == 'servers'
    insist { subject['server_name']   } == 'video-sunfire'

    # insist { subject['bytes'] }    == '19'
    # insist { subject['response'] } == '200'

    insist { subject['timestamp'] } == '01/May/2015:11:10:17.902'
    insist { subject['logsource'] } == 'haproxy-audio-lb-test'

    validate_haproxy_fields
  end

end

describe "haproxy filters" do

  extend LogStash::RSpec

  describe "with default config" do
    let(:parse_failure_tag) { '_grokparsefailure'}

    fields \
      'type'    => 'syslog',
      'program' => 'haproxy',
      'host'    => 'audio-lb-test'

    # type 'syslog'
    config [ 'filter{', File.read("conf.d/40_haproxy.conf"), '}' ].join

    it_behaves_like "a valid haproxy log parser"
  end

end


lib/logstash/inputs/s3.rb
# encoding: utf-8
require "logstash/inputs/base"
require "logstash/namespace"

require "time"
require "tmpdir"

# Stream events from files from a S3 bucket.
#
# Each line from each file generates an event.
# Files ending in '.gz' are handled as gzip'ed files.
class LogStash::Inputs::S3 < LogStash::Inputs::Base
  config_name "s3"
  milestone 1

  # TODO(sissel): refactor to use 'line' codec (requires removing both gzip
  # support and readline usage). Support gzip through a gzip codec! ;)
  default :codec, "plain"

  # The credentials of the AWS account used to access the bucket.
  # Credentials can be specified:
  # - As an ["id","secret"] array
  # - As a path to a file containing AWS_ACCESS_KEY_ID=... and AWS_SECRET_ACCESS_KEY=...
  # - In the environment (variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY)
  config :credentials, :validate => :array, :default => nil

  # The name of the S3 bucket.
  config :bucket, :validate => :string, :required => true

  # The AWS region for your bucket.
  config :region, :validate => ["us-east-1", "us-west-1", "us-west-2",
                                "eu-west-1", "ap-southeast-1", "ap-southeast-2",
                                "ap-northeast-1", "sa-east-1", "us-gov-west-1"],
                                :deprecated => "'region' has been deprecated in favor of 'region_endpoint'"

  # The AWS region for your bucket.
  config :region_endpoint, :validate => ["us-east-1", "us-west-1", "us-west-2",
                                "eu-west-1", "ap-southeast-1", "ap-southeast-2",
                                "ap-northeast-1", "sa-east-1", "us-gov-west-1"], :default => "us-east-1"

  # If specified, the prefix the filenames in the bucket must match (not a regexp)
  config :prefix, :validate => :string, :default => nil

  # Where to write the since database (keeps track of the date
  # the last handled file was added to S3). The default will write
  # sincedb files to some path matching "$HOME/.sincedb*"
  config :sincedb_path, :validate => :string, :default => nil

  # Name of a S3 bucket to backup processed files to.
  config :backup_to_bucket, :validate => :string, :default => nil

  # Path of a local directory to backup processed files to.
  config :backup_to_dir, :validate => :string, :default => nil

  # Whether to delete processed files from the original bucket.
  config :delete, :validate => :boolean, :default => false

  # Interval to wait between to check the file list again after a run is finished.
  # Value is in seconds.
  config :interval, :validate => :number, :default => 60

  # Max number of files to fetch from S3 in one iteration
  config :limit, :validate => :number, :default => 1000

  public
  def register
    require "digest/md5"
    require "aws-sdk"

    @region_endpoint = @region if !@region.empty?

    @logger.info("Registering s3 input", :bucket => @bucket, :region_endpoint => @region_endpoint)

    if @credentials.nil?
      @access_key_id = ENV['AWS_ACCESS_KEY_ID']
      @secret_access_key = ENV['AWS_SECRET_ACCESS_KEY']
    elsif @credentials.is_a? Array
      if @credentials.length ==1
        File.open(@credentials[0]) { |f| f.each do |line|
          unless (/^\#/.match(line))
            if(/\s*=\s*/.match(line))
              param, value = line.split('=', 2)
              param = param.chomp().strip()
              value = value.chomp().strip()
              if param.eql?('AWS_ACCESS_KEY_ID')
                @access_key_id = value
              elsif param.eql?('AWS_SECRET_ACCESS_KEY')
                @secret_access_key = value
              end
            end
          end
        end
        }
      elsif @credentials.length == 2
        @access_key_id = @credentials[0]
        @secret_access_key = @credentials[1]
      else
        raise ArgumentError.new('Credentials must be of the form "/path/to/file" or ["id", "secret"]')
      end
    end
    if @access_key_id.nil? or @secret_access_key.nil?
      raise ArgumentError.new('Missing AWS credentials')
    end

    if @bucket.nil?
      raise ArgumentError.new('Missing AWS bucket')
    end

    if @sincedb_path.nil?
      if ENV['HOME'].nil?
        raise ArgumentError.new('No HOME or sincedb_path set')
      end
      @sincedb_path = File.join(ENV["HOME"], ".sincedb_" + Digest::MD5.hexdigest("#{@bucket}+#{@prefix}"))
    end

    s3 = AWS::S3.new(
      :access_key_id => @access_key_id,
      :secret_access_key => @secret_access_key,
      :region => @region_endpoint
    )

    @s3bucket = s3.buckets[@bucket]

    unless @backup_to_bucket.nil?
      @backup_bucket = s3.buckets[@backup_to_bucket]
      unless @backup_bucket.exists?
        s3.buckets.create(@backup_to_bucket)
      end
    end

    unless @backup_to_dir.nil?
      Dir.mkdir(@backup_to_dir, 0700) unless File.exists?(@backup_to_dir)
    end

  end # def register

  public
  def run(queue)
    loop do
      process_new(queue)
      sleep(@interval)
    end
    finished
  end # def run

  private
  def process_new(queue, since=nil)

    if since.nil?
        since = sincedb_read()
    end

    while objects = list_new(since)
      break if objects.empty?
      objects.each_with_index do |k, i|
        @logger.debug("S3 input processing", :bucket => @bucket, :key => k)
        puts("#{Time.now} S3 input processing #{i} #{@bucket}/#{k}")
        lastmod = @s3bucket.objects[k].last_modified
        process_log(queue, k)
        sincedb_write(lastmod)
        since = sincedb_read()
      end
    end

  end # def process_new

  private
  def list_new(since=nil)

    if since.nil?
      since = Time.new(0)
    end

    i = 0
    objects = {}
    @s3bucket.objects.with_prefix(@prefix).each(:limit => @limit) do |log|
      puts "#{Time.now} #{i} #{log.key}"; i += 1
      if log.last_modified > since
        objects[log.key] = log.last_modified
      end
    end

    return sorted_objects = objects.keys.sort {|a,b| objects[a] <=> objects[b]}

  end # def list_new

  private
  def process_log(queue, key)

    object = @s3bucket.objects[key]
    tmp = Dir.mktmpdir("logstash-")
    begin
      filename = File.join(tmp, File.basename(key))
      File.open(filename, 'wb') do |s3file|
        object.read do |chunk|
          s3file.write(chunk)
        end
      end
      process_local_log(queue, filename)
      unless @backup_to_bucket.nil?
        backup_object = @backup_bucket.objects[key]
        backup_object.write(Pathname.new(filename))
      end
      unless @backup_to_dir.nil?
        FileUtils.cp(filename, @backup_to_dir)
      end
      if @delete
        object.delete()
      end
    end
    FileUtils.remove_entry_secure(tmp, force=true)

  end # def process_log

  private
  def process_local_log(queue, filename)

    metadata = {
      :version => nil,
      :format => nil,
    }
    File.open(filename) do |file|
      if filename.end_with?('.gz')
        gz = Zlib::GzipReader.new(file)
        gz.each_line do |line|
          metadata = process_line(queue, metadata, line)
        end
      else
        file.each do |line|
          metadata = process_line(queue, metadata, line)
        end
      end
    end

  end # def process_local_log

  private
  def process_line(queue, metadata, line)

    if /#Version: .+/.match(line)
      junk, version = line.strip().split(/#Version: (.+)/)
      unless version.nil?
        metadata[:version] = version
      end
    elsif /#Fields: .+/.match(line)
      junk, format = line.strip().split(/#Fields: (.+)/)
      unless format.nil?
        metadata[:format] = format
      end
    else
      @codec.decode(line) do |event|
        decorate(event)
        unless metadata[:version].nil?
          event["cloudfront_version"] = metadata[:version]
        end
        unless metadata[:format].nil?
          event["cloudfront_fields"] = metadata[:format]
        end
        queue << event
      end
    end
    return metadata

  end # def process_line

  private
  def sincedb_read()
    if File.exists?(@sincedb_path)
      content = ::File.read(@sincedb_path).chomp.strip
      # If the file was created but we didn't have the time to write to it
      return content.empty? ? Time.new(0) : Time.parse(content)
    else
      return Time.new(0)
    end

  end # def sincedb_read

  private
  def sincedb_write(since=nil)

    if since.nil?
      since = Time.now()
    end
    File.open(@sincedb_path, 'w') { |file| file.write(since.to_s) }
  end # def sincedb_write

end # class LogStash::Inputs::S3


lib/logstash/filters/s3_access_log.rb
# encoding: utf-8

#
# This filter will convert  Amazon S3 Server Access Log format to
# Apache Combined Log Format (CLF).
#

require "logstash/filters/base"
require "logstash/namespace"

#
# This filter will convert  Amazon S3 Server Access Log format to
# Apache Combined Log Format (CLF).
#
# Amazon S3 Server Access Log Format is documented at:
#   http://docs.aws.amazon.com/AmazonS3/latest/dev/LogFormat.html
#
# Apache Combined Log Format is documented at:
#   https://httpd.apache.org/docs/trunk/logs.html#combined
#
# The config should look like this:
#
#     filter {
#       s3_access_log {
#         copy_operation => "drop"  # "drop" or "convert" ("convert" the default))
#       }
#     }
#
# The 'copy_operation' option controls how Amazon Additional Logging for Copy Operations
# (REST.COPY.OBJECT_GET) is handled.
# See: http://docs.aws.amazon.com/AmazonS3/latest/dev/LogFormat.html#AdditionalLoggingforCopyOperations
#
# An empty request_uri with REST.COPY_OBJECT_GET operations is
# replaced with a request_uri 'POST /<key>' with referrer "REST.COPY.OBJECT_GET"
#
class LogStash::Filters::S3AccessLog < LogStash::Filters::Base

  # convert the format as specified in http://docs.aws.amazon.com/AmazonS3/latest/dev/LogFormat.html
  #
  # 1 Bucket Owner
  #   The canonical user id of the owner of the source bucket.
  # 2 Bucket
  #   The name of the bucket that the request was processed against. If the system receives a malformed request and cannot determine the bucket, the request will not appear in any server access log.
  # 3 Time
  #   The time at which the request was received. The format, using strftime() terminology, is [%d/%b/%Y:%H:%M:%S %z]
  # 4 Remote IP
  #   The apparent Internet address of the requester. Intermediate proxies and firewalls might obscure the actual address of the machine making the request.
  # 5 Requester
  #   The canonical user id of the requester, or the string "Anonymous" for unauthenticated requests. This identifier is the same one used for access control purposes.
  # 6 Request ID
  #   The request ID is a string generated by Amazon S3 to uniquely identify each request.
  # 7 Operation
  #   Either SOAP.operation, REST.HTTP_method.resource_type or WEBSITE.HTTP_method.resource_type
  # 8 Key
  #   The "key" part of the request, URL encoded, or "-" if the operation does not take a key parameter.
  # 9 Request-URI
  #   The Request-URI part of the HTTP request message.
  # 10 HTTP status
  #   The numeric HTTP status code of the response.
  # 11 Error Code
  #   The Amazon S3 Error Code, or "-" if no error occurred.
  # 12 Bytes Sent
  #   The number of response bytes sent, excluding HTTP protocol overhead, or "-" if zero.
  # 13 Object Size
  #   The total size of the object in question.
  # 14 Total Time
  #   The number of milliseconds the request was in flight from the server's perspective. This value is measured from the time your request is received to the time that the last byte of the response is sent. Measurements made from the client's perspective might be longer due to network latency.
  # 15 Turn-Around Time
  #   The number of milliseconds that Amazon S3 spent processing your request. This value is measured from the time the last byte of your request was received until the time the first byte of the response was sent.
  # 16 Referrer
  #   The value of the HTTP Referrer header, if present. HTTP user-agents (e.g. browsers) typically set this header to the URL of the linking or embedding page when making a request.
  # 17 User-Agent
  #   The value of the HTTP User-Agent header.
  # 18 Version Id
  #   The version ID in the request, or "-" if the operation does not take a versionId parameter.
  class S3AccessLogLine

    # Amazon S3 Server Access Log Format is documented at:
    #   http://docs.aws.amazon.com/AmazonS3/latest/dev/LogFormat.html
    AMAZON_S3_ACCESS_LOG_FORMAT  = Regexp.new('([^ ]*) ([^ ]*) \[([^\]]*)\] ([^ ]*) ([^ ]*) ([^ ]*) ([^ ]*) ([^ ]*) ("[^"]*"|-) ([^ ]*|-) ([^ ]*|-) ([^ ]*|-) ([^ ]*|-) ([^ ]*|-) ([^ ]*|-) ("[^"]*"|-) ("[^"]*"|-) ([^ ]*|-)( ?.*)')

    # Parse 'GET /11625060-v901740/20140907093122_J1084854-mp4.mp4 HTTP/1.1' into %w(GET /11625060-v901740/20140907093122_J1084854-mp4.mp4 1.1)
    AMAZON_S3_REQUEST_URI_FORMAT = Regexp.new('(\b\w+\b) (\S+)(?: HTTP\/(\d+\.\d+))?')

    S3_COPY_OPERATION = 'REST.COPY.OBJECT_GET'.freeze

    APACHE_CLF_FORMAT = "%s - %s [%s] %s %s %s %s %s %d".freeze

    def initialize(event, source_field)
      @source_line = event[source_field]
      if @source_line =~ AMAZON_S3_ACCESS_LOG_FORMAT

        event['owner']              = $1
        event['bucket']             = $2
        event['timestamp']          = $3
        event['remote_ip']          = $4
        event['requester']          = $5
        event['request_id']         = $6
        event['operation']          = $7
        event['key']                = '-' == $8 ? nil : $8 # remove empty
        event['request_uri']        = $9
        event['http_status']        = $10
        event['error_code']         = '-' == $11 ? nil : $11 # remove empty
        event['bytes']              = '-' == $12 ? 0   : $12 # bytes instead of bytes_sent intentional
        event['object_size']        = $13
        event['total_time_ms']      = $14
        event['turnaround_time_ms'] = $15
        event['referrer']           = $16
        event['agent']              = $17 # agent instead of user_agent intentional
        event['version_id']         = '-' == $18 ? nil : $18 # remove empty
        event['trailing_fields']    = $19.strip! rescue nil  # trailing fields as single string or nil if empty

        # disect request_uri
        if event['request_uri'] =~ AMAZON_S3_REQUEST_URI_FORMAT
          event['verb'], event['request'], event['httpversion'] = $1, $2, $3
        else
          event['rawrequest'] = event['request_uri']
        end
      else
        ArgumentError.new("Line not in Amazon S3 access log format: #{@source_line}")
      end
    end

    #
    # Handle Additional Logging for Copy Operations (REST.COPY.OBJECT_GET)
    # See: http://docs.aws.amazon.com/AmazonS3/latest/dev/LogFormat.html#AdditionalLoggingforCopyOperations
    #
    # An empty request_uri with REST.COPY_OBJECT_GET operations is
    # replaced with a request_uri 'POST /<key>' with referrer "REST.COPY.OBJECT_GET"
    #
    def convert_copy_operation!(event)
      raise ArgumentError.new("Not a #{S3_COPY_OPERATION}: #{@source_line}") unless is_copy_operation?(event)
      event['referrer']    = %("#{event['operation']}")
      event['agent']       = '"-"'
      event['request_uri'] = %("POST /#{event['key']} HTTP/1.1") # output as POST
      event['bytes']       = '0'
    end

    def is_copy_operation?(event)
      '-' == event['request_uri'] && S3_COPY_OPERATION == event['operation']
    end

    #
    # 206 Partial Content requests can have excessive value for bytes_sent because S3 registers the
    # bytes pushed onto the network.
    # This method recalculates the bytes_sent to estimate the bytes received by the client device
    # based on an assumed bitrate of 24 kbit/sec.
    #
    # Assume 128 Kbytes buffer is ingested and total_time_ms
    #
    def recalculate_partial_content!(event, max_kbitrate)
      if event['key'] =~ /\.mp3$/
        if 206 == event['http_status'].to_i && ((event['bytes'].to_i*8)/event['total_time_ms'].to_i > 2000)
          event['bytes'] = [ 128 * 1024 + ((max_kbitrate/8000.0).round) * event['total_time_ms'].to_i, event['bytes'].to_i ].min # 128 K buffer + 3 bytes/msec = 3 kbytes/sec = 24 kbit/sec
          event.tag RECALCULATED_TAG
        end
      end
    end

    # Apache Combined Log Format is documented at:
    #   https://httpd.apache.org/docs/trunk/logs.html#combined
    def to_apache_clf(event)
      APACHE_CLF_FORMAT % [ event['remote_ip'],   requester(event), event['timestamp'], event['request_uri'],
                            event['http_status'], event['bytes'],   event['referrer'],  event['agent'], duration(event) ]
    end

    private

    # shorten requester to max 10 characters
    def requester(event)
      event['requester'][0..9]
    end

    def duration(event)
      (event['total_time_ms'].to_i / 1000.0).round
    end

  end

  config_name "s3_access_log"
  milestone 1

  # The field to convert to JSON.
  config :source, :validate => :string

  # The field to write the JSON into. If not specified, the source
  # field will be overwritten.
  config :target, :validate => :string

  # Config option to control how to handle REST.COPY.OBJECT_GET lines.
  # Either drop or convert to POST requests.
  # See: http://docs.aws.amazon.com/AmazonS3/latest/dev/LogFormat.html#AdditionalLoggingforCopyOperations
  config :copy_operation, :validate => [ 'convert', 'drop' ], :default => 'convert'

  # Config option to control whether to recalculate 206 Partial Content requests to
  # calculate reasonable byte counts. Don't pass cost of buffering at S3 to customers.
  config :recalculate_partial_content, :validate => :boolean, :default => false

  config :max_kbitrate, :validate => :number, :default => 24000

  PARSE_FAILURE_TAG = "_s3parsefailure".freeze
  RECALCULATED_TAG = "bytes_recalculated".freeze

  public
  def register
    @source ||= 'message'
    @target ||= 'message'
  end # def register

  public
  def filter(event)
    return unless filter?(event)

    @logger.debug("Running S3 Server Access Log filter", :event => event)

    line = S3AccessLogLine.new(event, @source)

    if line.is_copy_operation?(event)
      case @copy_operation
      when 'drop'
        event.cancel
        return
      when 'convert'
        line.convert_copy_operation!(event)
      else
        raise ArgumentError.new("Invalid copy operation: #{@copy_operation}")
      end
    end

    line.recalculate_partial_content!(event, @max_kbitrate) if @recalculate_partial_content

    event[@target] = line.to_apache_clf(event)
    filter_matched(event)

    @logger.debug? && @logger.debug("Event after S3AccessLog filter", :event => event)

  rescue => e
    event.tag PARSE_FAILURE_TAG
    @logger.warn("Exception for S3 access log", :source => @source, :raw => event[@source].inspect, :exception => e)
  end # def filter

end # class LogStash::Filters::S3AccessLog
